index.js

```diff
@@ -32,54 +33,105 @@ load.path = function (dir) {
+  var prebuilds = path.join(dir, 'prebuilds', platform + '-' + arch)
+  var parsed = readdirSync(prebuilds).map(parseTags)
+  var candidates = parsed.filter(matchTags(runtime, abi))
+  var winner = candidates.sort(compareTags(runtime))[0]
+  if (winner) return path.join(prebuilds, winner.file)
+
+  var target = [
+    'platform=' + platform,
+    'arch=' + arch,
+    'runtime=' + runtime,
+    'abi=' + abi,
+    'uv=' + uv,
+    armv ? 'armv=' + armv : '',
+    'libc=' + libc
+  ].filter(Boolean).join(' ')
+
+  throw new Error('No native build was found for ' + target)
 }
 
-function getFirst (dir, filter) {
+function readdirSync (dir) {
   try {
-    var files = fs.readdirSync(dir).filter(filter)
-    return files[0] && path.join(dir, files[0])
+    return fs.readdirSync(dir)
   } catch (err) {
-    return null
+    return []
   }
 }
 
-function matchNapiRuntime (name) {
-  return name === runtime + '-napi.node'
+function getFirst (dir, filter) {
+  var files = readdirSync(dir).filter(filter)
+  return files[0] && path.join(dir, files[0])
 }
 
-function matchNapi (name) {
-  return name === 'node-napi.node'
+function matchBuild (name) {
+  return /\.node$/.test(name)
 }
 
-function matchPrebuild (name) {
-  var parts = name.split('-')
-  return parts[0] === runtime && parts[1] === abi + '.node'
+function parseTags (file) {
+  var arr = file.split('.')
+  var extension = arr.pop()
+  var tags = { file: file, specificity: 0 }
+
+  if (extension !== 'node') return
+
+  for (var i = 0; i < arr.length; i++) {
+    var tag = arr[i]
+
+    if (tag === 'node' || tag === 'electron' || tag === 'node-webkit') {
+      tags.runtime = tag
+    } else if (tag === 'napi') {
+      tags.napi = true
+    } else if (tag.slice(0, 3) === 'abi') {
+      tags.abi = tag.slice(3)
+    } else if (tag.slice(0, 2) === 'uv') {
+      tags.uv = tag.slice(2)
+    } else if (tag.slice(0, 4) === 'armv') {
+      tags.armv = tag.slice(4)
+    } else if (tag === 'glibc' || tag === 'musl') {
+      tags.libc = tag
+    } else {
+      continue
+    }
+
+    tags.specificity++
+  }
+
+  return tags
 }
 
-function matchBuild (name) {
-  return /\.node$/.test(name)
+function matchTags (runtime, abi) {
+  return function (tags) {
+    if (tags == null) return false
+    if (tags.runtime !== runtime && !runtimeAgnostic(tags)) return false
+    if (tags.abi !== abi && !tags.napi) return false
+    if (tags.uv && tags.uv !== uv) return false
+    if (tags.armv && tags.armv !== armv) return false
+    if (tags.libc && tags.libc !== libc) return false
+
+    return true
+  }
+}
+
+function runtimeAgnostic (tags) {
+  return tags.runtime === 'node' && tags.napi
+}
+
+function compareTags (runtime) {
+  // Precedence: non-agnostic runtime, abi over napi, then by specificity.
+  return function (a, b) {
+    if (a.runtime !== b.runtime) {
+      return a.runtime === runtime ? -1 : 1
+    } else if (a.abi !== b.abi) {
+      return a.abi ? -1 : 1
```

Comment:
@vweevers Mind explaining this a bit?

Reply:
We want to pick a prebuild that most closely matches the runtime (environment).

By this point in the code, we have already filtered out prebuilds with non-matching tags. But let's say the current runtime is Electron. There could still be:

1. A prebuild specifically targeting Electron
2. A runtime-agnostic prebuild (suitable for both Node and Electron)

If (1) exists, it should sort first and win. We're assuming that if (1) exists, it must mean that (2), though agnostic for _most_ runtimes, is not suitable for this specific runtime.

The logic is the same for the ABI's. At this point in the code, we could have:

1. A prebuild specifically targeting the current ABI
2. An "ABI-agnostic" prebuild (i.e. N-API)

If (1) exists, it should sort first and win.